services/pull/pull.go

```diff
@@ -371,15 +373,46 @@ func checkIfPRContentChanged(ctx context.Context, pr *issues_model.PullRequest,
 		return false, fmt.Errorf("GetMergeBase: %w", err)
 	}
 
-	cmd := git.NewCommand(ctx, "diff", "--name-only", "-1").AddDynamicArguments(newCommitID, oldCommitID, base)
-	stdout, stderr, err := cmd.RunStdString(&git.RunOpts{
-		Dir: tmpBasePath,
-	})
+	cmd := git.NewCommand(ctx, "diff", "--name-only").AddDynamicArguments(newCommitID, oldCommitID, base)
+	stdoutReader, stdoutWriter, err := os.Pipe()
 	if err != nil {
-		log.Error("Unable to run diff on %s %s %s in tempRepo for PR[%d]%s/%s...%s/%s: stdout %s stderr %s Error: %v",
+		return false, fmt.Errorf("unable to open pipe for to run diff: %w", err)
+	}
+
+	changedErr := fmt.Errorf("changed files")
+
+	if err := cmd.Run(&git.RunOpts{
+		Dir:    tmpBasePath,
+		Stdout: stdoutWriter,
+		PipelineFunc: func(ctx context.Context, cancel context.CancelFunc) error {
+			_ = stdoutWriter.Close()
+			defer func() {
+				_ = stdoutReader.Close()
+			}()
+			buf := make([]byte, 1024)
+			for {
+				n, err := stdoutReader.Read(buf)
+				if err != nil {
+					if err == io.EOF {
+						return nil
+					}
+					return err
+				}
+
+				ts := bytes.TrimSpace(buf[:n])
+				if len(ts) > 0 {
+					return changedErr
+				}
+			}
+		},
```

Comment:
That is very silly.

1. This function has no further reuse.
2. It makes the code less clear as to what it does.
3. It would require making changedErr a const

Reply:
1. Gitea is pretty much the only application I know that regularly has methods with a length of 150+ lines.
While about two thirds of it are due to the design of Go, that is still way too much.
There are parts in the codebase no one wants to look at simply because they are too complicated.
2. I don't think giving it an explicit name makes it less clear.
3. Okay, yeah, I overlooked the ` return changedErr`.
What we could do instead is extract the body of this function, and give it an additional param `customErr` which we return in this case.
Then, we can call this function with the two arguments + the customErr inside the anonymous pipelinefunc.

Reply:
I generally agree that long functions are a big problem in Gitea - and you should be aware that I have been one of the most active in reducing the size of functions in Gitea.

HOWEVER, every function you separate out is a context switch unless you can give it a very clear name. 

The original code does not have a good name - but through reading the c code for git diff I can see that I can make a nicer function with a clear name, however, the pipeline function is still not completely extracted and could only be extracted using a function that returns a function.